articles/avoid-fix-flaky-tests.md

```diff
@@ -0,0 +1,73 @@
+If you have to deal with flaky tests that fail very rarely, running the tests a few times a day might not give you enough data for analysis. Therefore schedule the tests to run as often as you can afford. In a recent project we ran the test suite once every hour throughout the night. This gave us a lot of data to analyze stability the next morning.
+
+## Aggregate Results From Feature Branches
+
+You might actually have a lot of data from running the tests on each feature branch, but it is hard to see patterns if you don't have the test runs in an aggregated form. Collect all the test failures from all branches in a single place to analyze them. You might start off with an Excel sheet, but will probably automate the data collection soon.
+
+## Visualize
+
+Visualize failures in a matrix where one dimension is the test run and the other is the test scenario (like the [Jenkins Test Results Analyzer Plugin](https://wiki.jenkins.io/display/JENKINS/Test+Results+Analyzer+Plugin) does. This might reveal patterns that can give you a hint where the problem lies. E.g. if a certain group of tests always fail together, you might want to check what they have in common. And if a single test only fails every 10th time or so, painting it on a graph will make it apparent that there's a problem with this test.
+
+## Analyze Logs and Metrics
+
+Logs and metrics from the CI server that runs the test and from the SUT might give you valuable hints to the root cause. If your servers run in VMs, you might also want to collect metrics and log files from the VM host. Collect everything in a central place so you can analyze the timeline of events and look for reoccurring patterns.
+
+
+# If All Else Fails
+
+## Rerun Failed Tests Automatically
+
+If you can't tame the beast or if you need a quick fix to buy you some time for a proper solution you might want to rerun failed tests once or twice automatically (in Java you could use a JUnit rule for this). If the reruns is successful, the build is successful. This will make your build runs a bit slower, but might spare you a rerun of the entire build pipeline.
+
+## Rethink Your Automation Strategy
+
+Your test strategy is good if you can't find a cheaper approach that gives you a similar confidence. Think economically. How much time do you spend fixing the brittle tests? How many bugs do you find using these tests? How much time do you save because of the quick feedback you get? How much exploratory testing can you save because of your automated tests? Does the benefit of the tests pay for the bill of maintaining them?
+
+If, despite all your efforts, you find it impossible to get the test suite stable, rethink your automation strategy. Is it really necessary to program tests so high up in the testing pyramid? Do you get enought value from the flaky tests so it's worth keeping them? Or might it be more economic to delete the problematic tests and replace them with lower level automated tests plus more exploratory testing?
+
```

Comment:
This paragraph seems to repeat of the questions from the previous one. Maybe you could transform them to statements. e.g

It might no be necessary to have all tests on the top of the testing pyramid. If you can't get enough value from the flaky tests, it might no be worth keeping (all of) them. Some tests could be replaced with lower level automated tests and covered by additional exploratory testing.

Reply:
Rewriting this paragraph